src/DomainHost.sol

```diff
@@ -0,0 +1,490 @@
+            vat.frob(ilk, address(this), address(this), address(this), diff, diff);
+            daiJoin.exit(escrow, uint256(diff));
+            grain = _grain;
+        }
+
+        dai.transferFrom(address(escrow), address(this), _wad);
+        daiJoin.join(address(vow), _wad);
+
+        ddai = 0;
+
+        emit Accrue(_wad);
+    }
+
+    /// @notice Move bad debt from the remote domain into the local vow
+    /// @dev This is a potentially dangerous operation as a malicious domain can drain the entire surplus buffer
+    /// Because of this we require an authed party to perform this operation
+    function _rectify() internal auth returns (uint256 _rid, uint256 _wad) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+        uint256 _dsin = dsin;
+        uint256 _ddai = ddai;
+        require(_dsin > _ddai, "DomainHost/no-deficit");
+
+        unchecked { _wad = _dsin - _ddai; }
+        vat.suck(vow, address(this), _wad * RAY);
+        daiJoin.exit(address(escrow), _wad);
+        dsin = 0;
+        ddai = 0;
+
+        _rid = rid++;
+
+        emit Rectify(_wad);
+    }
+
+    /// @notice Initiate shutdown for this domain
+    /// @dev This will trigger the end module on the remote domain
+    function _cage() internal returns (uint256 _rid) {
+        require(vat.live() == 0 || wards[msg.sender] == 1, "DomainHost/not-authorized");
+        require(live == 1, "DomainHost/not-live");
+
+        live = 0;
+
+        _rid = rid++;
+
+        emit Cage();
+    }
+
+    /// @notice Process this domain debt
+    /// @param _lid Local ordering id
+    /// @param _debt The value of debt [RAD]
+    function _tell(uint256 _lid, uint256 _debt) internal ordered(_lid) {
+        require(live == 0, "DomainHost/live");
+        require(!debtReported, "DomainHost/debt-reported");
+
+        uint256 wad = grain + ddai;
+        uint256 debtW = _divup(_debt, RAY);
+        if (wad > debtW) {
+            unchecked {
+                wad = wad - debtW;
+            }
+            dai.transferFrom(address(escrow), address(this), wad);
+            daiJoin.join(address(vow), wad);
+        }
+
+        debt = _debt;
+        debtReported = true;
+
+        emit Tell(_debt);
+    }
+
+    /// @notice Allow DAI holders to exit during global settlement
+    /// @param usr The address to send the claim token to
+    /// @param wad The amount of gems to exit [WAD]
+    function _exit(address usr, uint256 wad) internal returns (uint256 claim) {
+        require(vat.live() == 0, "DomainHost/vat-live");
+        vat.slip(ilk, msg.sender, -_int256(wad));
+        claim = wad * (debt / grain); // Round against the user
+
+        emit Exit(msg.sender, usr, wad, claim);
+    }
+
+    /// @notice Allow DAI holders to exit during global settlement
+    /// @param usr The address to send the claim token to
+    /// @param wad The amount of gems to exit [WAD]
+    function _exit(bytes32 usr, uint256 wad) internal returns (uint256 claim) {
```

Comment:
This seems to be a copy of `_exit` from line 395?

Reply:
hmmm no, this one is expecting a `bytes32` `usr` instead of the standard `address`. I'd assume this is the one you guys would use.

Reply:
excuse my dyslexia